File: dotnet/Allmark/Block/ListBulletedRule.cs
namespace Allmark.Block;

using Allmark.Types;

public static class ListBulletedRule
{
    public static BlockRule Create()
    {
        return new BlockRule
        {
            Name = "list_bulleted",
            TestStart = TestStart,
            TestContinue = TestContinue,
            CloseNode = CloseNode,
        };
    }

    private static ListInfo? GetMarkup(BlockParserState state)
    {
        var ch = Utils.GetChar(state.Src, state.I);
        if (
            (ch == '-' || ch == '+' || ch == '*') &&
            (state.I == state.Src.Length - 1 || Utils.IsSpace(state.Src[state.I + 1])))
        {
            var delimiter = Utils.GetChar(state.Src, state.I).ToString();
            var nextChar = Utils.GetChar(state.Src, state.I + 1);
            return new ListInfo
            {
                Delimiter = delimiter,
                Markup = delimiter,
                IsBlank = state.I == state.Src.Length - 1 || nextChar == '\n' || nextChar == '\r',
                Type = "list_bulleted"
            };
        }
        return null;
    }

    private static bool TestStart(BlockParserState state, MarkdownNode parent, int endOfLine)
    {
        if (parent.AcceptsContent)
        {
            return false;
        }

        var info = GetMarkup(state);
        return ListRule.TestListStart(state, parent, endOfLine, info);
    }

    private static bool TestContinue(BlockParserState state, MarkdownNode node)
    {
        var info = GetMarkup(state);
        return ListRule.TestListContinue(state, node, info);
    }

    private static void CloseNode(BlockParserState state, MarkdownNode node)
    {
        node.Loose = ListRule.IsLooseList(node);
    }
}

File: zig/src/block/listBulletedRule.zig
const std = @import("std");
const BlockParserState = @import("../types/BlockParserState.zig").BlockParserState;
const BlockRule = @import("../types/BlockRule.zig").BlockRule;
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const isNumeric = @import("../utils/isAlphaNumeric.zig").isNumeric;
const isNewLine = @import("../utils/isNewLine.zig").isNewLine;
const isSpace = @import("../utils/isSpace.zig").isSpace;
const getMarkup = @import("./listRule.zig").getMarkup;
const testListStart = @import("./listRule.zig").testListStart;
const testListContinue = @import("./listRule.zig").testListContinue;
const isLooseList = @import("./listRule.zig").isLooseList;
const movePastMarker = @import("../utils/movePastMarker.zig").movePastMarker;

pub fn testStart(state: *BlockParserState, parent: *MarkdownNode, end_of_line: usize) bool {
    if (parent.acceptsContent) {
        return false;
    }

    const info = getMarkup(state);
    if (info == null) return false;

    return testListStart(state, parent, end_of_line, info);
}

pub fn testContinue(state: *BlockParserState, node: *MarkdownNode) bool {
    const info = getMarkup(state);
    return testListContinue(state, node, info);
}

fn closeNode(state: *BlockParserState, node: *MarkdownNode) void {
    _ = state;
    node.loose = isLooseList(node);
}

pub const listBulletedRule = BlockRule{
    .name = "list_bulleted",
    .testStart = testStart,
    .testContinue = testContinue,
    .closeNode = closeNode,
};
